src/raft/raft.go

```diff
@@ -41,25 +46,319 @@ type ApplyMsg struct {
+			return
+		case vote := <-votes:
+			if vote.Term > rf.getTerm() {
+				rf.makeFollower()
+				rf.setTerm(vote.Term)
+			}
+			if vote.VoteGranted {
+				tally++
+			}
+			if tally >= quorum {
+				DPrintf("👺 I'm the boss %v", rf.me)
+
+				rf.makeLeader()
+				return
+			}
+		}
+	}
+}
+
+func (rf *Raft) dispatchVotes() (votes chan *RequestVoteReply) {
+	// Vote for yourself
+	others := exclude(rf.peers, rf.me)
+	votes = make(chan *RequestVoteReply, len(others))
+	votes <- &RequestVoteReply{rf.getTerm(), true}
+	request := rf.makeVoteRequest()
+	for _, v := range others {
+		go func(v int) {
+			DPrintf("Requesting vote with: %+v", request)
+			reply := &RequestVoteReply{}
+			rf.sendRequestVote(v, request, reply)
+			votes <- reply
+		}(v)
+	}
+	return votes
+}
+
+func (rf *Raft) makeVoteRequest() RequestVoteArgs {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	request := RequestVoteArgs{rf.getTerm(), rf.getID()}
+
+	DPrintf("Request constructed: %+v", request)
+
+	return request
+}
+
+func (rf *Raft) makeAppendEntries(arg string) AppendEntriesArgs {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	request := AppendEntriesArgs{rf.getTerm(), rf.getID()}
+	return request
+}
+
+func (rf *Raft) dispathAppendEntries() (replies chan *AppendEntriesReply) {
+	DPrintf("⏲  💖 Leader %v sent appendEntriesRPC", rf.me)
+	others := exclude(rf.peers, rf.me)
+	DPrintf("Others: %v", others)
+	replies = make(chan *AppendEntriesReply, len(others))
+	request := rf.makeAppendEntries("Ping")
+	for _, peer := range others {
+		go func(v int) {
+			DPrintf("Sending to %v Appending with: %+v", v, request)
+			reply := &AppendEntriesReply{}
+			now := time.Now()
+			ok := rf.sendAppendEntries(v, request, reply)
+			if ok {
+				diff := time.Since(now)
+				DPrintf("Request took, %+v", diff)
+			}
+			replies <- reply
+		}(peer)
+	}
+	return replies
+}
+
+func (rf *Raft) runLeader() {
+	heartBeat := randomTimeout(rf.heartbeatTimeout)
+	results := rf.dispathAppendEntries()
+	quorum := rf.getMajority()
+	contact := 1
+
+	for {
+		select {
+		case <-rf.killCh:
+			close(rf.killCh)
+			DPrintf("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx------------------  Killed ")
+			return
+		case reply := <-results:
+			DPrintf("Got something at %v", time.Now())
+			rf.setLastContact()
+			if reply.Term > rf.getTerm() {
+				DPrintf("Greater Term")
+				rf.setTerm(reply.Term)
+				rf.makeFollower()
+				return
+			}
+			contact++
+		case <-heartBeat:
+			if delay := time.Since(rf.lastContact); delay > (2000 * time.Millisecond) {
+				DPrintf("Took too long. Delay was %v. Timeout is: %v", delay, rf.heartbeatTimeout)
+				rf.makeFollower()
+				return
+			}
+			if contact < quorum {
+				DPrintf("Not enough replies")
+				rf.makeFollower()
+				return
+			}
+			results = rf.dispathAppendEntries()
+			heartBeat = randomTimeout(rf.heartbeatTimeout)
+		case <-rf.heartBeatChan:
+			DPrintf("💖 Candidate %v received heartbeat", rf.me)
+			rf.makeFollower()
+			return
+		}
+	}
```

Comment:
Leaders send out heartbeat messages to their followers to maintain a link. If they lose contact for too long, get messages from other leaders, or don't reach enough followers they react to that appropriately.